crates/base/src/rt_worker/supervisor/strategy_per_request.rs

```diff
@@ -0,0 +1,118 @@
+use std::time::Duration;
+
+use event_worker::events::ShutdownReason;
+use log::error;
+use sb_workers::context::{Timing, UserWorkerMsgs};
+use tokio::time::Instant;
+
+use crate::rt_worker::supervisor::{handle_interrupt, IsolateInterruptData};
+
+use super::Arguments;
+
+pub async fn supervise(args: Arguments, oneshot: bool) -> ShutdownReason {
+    let Arguments {
+        key,
+        runtime_opts,
+        timing,
+        mut cpu_alarms_rx,
+        mut memory_limit_rx,
+        pool_msg_tx,
+        isolate_memory_usage_tx,
+        thread_safe_handle,
+        cpu_timer,
+        ..
+    } = args;
+
+    let Timing {
+        req: (mut req_start_rx, mut req_end_rx),
+        ..
+    } = timing.unwrap_or_default();
+
+    let mut complete_reason = None::<ShutdownReason>;
+    let mut req_start_ack = false;
+
+    // reduce 100ms from wall clock duration, so the interrupt can be handled before
+    // isolate is dropped
+    let wall_clock_duration =
+        Duration::from_millis(runtime_opts.worker_timeout_ms) - Duration::from_millis(100);
+
+    let wall_clock_duration_alert = tokio::time::sleep(wall_clock_duration);
+
+    tokio::pin!(wall_clock_duration_alert);
+    loop {
+        tokio::select! {
+            Some(_) = cpu_alarms_rx.recv() => {
+                if req_start_ack {
+                    error!("CPU time limit reached. isolate: {:?}", key);
+                    complete_reason = Some(ShutdownReason::CPUTime);
+                }
+            }
+
+            Some(notify) = req_start_rx.recv() => {
+                // INVARIANT: This branch MUST not be satisfied more than once
+                // during the same request cycle.
+                assert!(!req_start_ack, "supervisor has seen request start signal twice");
+
+                notify.notify_one();
+
+                if let Some(x) = cpu_timer.as_ref() {
```

Comment:
Awesome PR! Could we have a more descriptive name for `x` ?

Reply:
>

> How's this line working also? I was getting an error, because `reset` is not part of CPUTimer, because it does not contain `&self`.
> 
> had to add
> 
> ```rust
> #[cfg(not(target_os = "linux"))]
>     pub fn reset(&self) -> Result<(), Error> {
>         Ok(())
>     }
> ```
> 
> (&self) for it to compile

Ah, that was probably my mistake.
As for the CPU timer, it's only available on Linux, so I hadn't actively looked at other platforms behind the feature flag.

I'll fix it

Reply:
solved